gfx_direct3d11.cpp

```diff
@@ -0,0 +1,976 @@
+    swap_chain_description.SampleDesc = d3d.sample_description;
+    swap_chain_description.SwapEffect = DXGI_SWAP_EFFECT_DISCARD;
+    swap_chain_description.Windowed = TRUE;
+
+    // Create device and swap chain
+
+#if DEBUG_D3D
+    UINT device_creation_flags = D3D11_CREATE_DEVICE_DEBUG;
+#else
+    UINT device_creation_flags = 0;
+#endif
+
+    ThrowIfFailed(D3D11CreateDeviceAndSwapChain(
+        nullptr,
+        D3D_DRIVER_TYPE_HARDWARE,
+        nullptr,
+        device_creation_flags,
+        nullptr,
+        0,
+        D3D11_SDK_VERSION,
+        &swap_chain_description,
+        d3d.swap_chain.GetAddressOf(),
+        d3d.device.GetAddressOf(),
+        nullptr,
+        d3d.context.GetAddressOf()));
+
+#if DEBUG_D3D
+    ThrowIfFailed(d3d.device->QueryInterface(__uuidof(ID3D11Debug), (void **) (d3d.debug.GetAddressOf())));
+#endif
+
+    // Create views
+
+    create_render_target_views(DESIRED_SCREEN_WIDTH, DESIRED_SCREEN_HEIGHT);
+
+    // Create main vertex buffer
+
+    D3D11_BUFFER_DESC vertex_buffer_desc;
+    ZeroMemory(&vertex_buffer_desc, sizeof(D3D11_BUFFER_DESC));
+
+    vertex_buffer_desc.Usage = D3D11_USAGE_DYNAMIC;
+    vertex_buffer_desc.ByteWidth = 256 * 26 * 3 * sizeof(float); // Same as buf_vbo size in gfx_pc
+    vertex_buffer_desc.BindFlags = D3D11_BIND_VERTEX_BUFFER;
+    vertex_buffer_desc.CPUAccessFlags = D3D11_CPU_ACCESS_WRITE;
+    vertex_buffer_desc.MiscFlags = 0;
+
+    ThrowIfFailed(d3d.device->CreateBuffer(&vertex_buffer_desc, NULL, d3d.vertex_buffer.GetAddressOf()));
+
+    // Create constant buffer
+
+    D3D11_BUFFER_DESC constant_buffer_desc;
+    ZeroMemory(&constant_buffer_desc, sizeof(D3D11_BUFFER_DESC));
+
+    constant_buffer_desc.Usage = D3D11_USAGE_DYNAMIC;
+    constant_buffer_desc.ByteWidth = sizeof(PerFrameCB);
+    constant_buffer_desc.BindFlags = D3D11_BIND_CONSTANT_BUFFER;
+    constant_buffer_desc.CPUAccessFlags = D3D11_CPU_ACCESS_WRITE;
+    constant_buffer_desc.MiscFlags = 0;
+
+    ThrowIfFailed(d3d.device->CreateBuffer(&constant_buffer_desc, NULL, d3d.per_frame_cb.GetAddressOf()));
+
+    // Show the window
+
+    QueryPerformanceFrequency(&frequency);
+    QueryPerformanceCounter(&last_time);
+
+    ShowWindow(h_wnd, SW_SHOW);
+    UpdateWindow(h_wnd);
+}
+
+static void gfx_d3d11_dxgi_main_loop(void (*run_one_game_iter)(void)) {
+    MSG msg = { 0 };
+
+    if (PeekMessage(&msg, NULL, 0, 0, PM_REMOVE)) {
+        TranslateMessage(&msg);
+        DispatchMessage(&msg);
+    } else {
+        if (IsIconic(h_wnd)) {
+            Sleep(1);
+            return;
+        }
+
+        LARGE_INTEGER current_time, elapsed_time_microseconds;
+        QueryPerformanceCounter(&current_time);
+        elapsed_time_microseconds.QuadPart = current_time.QuadPart - last_time.QuadPart;
+        elapsed_time_microseconds.QuadPart *= 1000000;
+        elapsed_time_microseconds.QuadPart /= frequency.QuadPart;
+
+        if (elapsed_time_microseconds.QuadPart < 33333.33333333) {
```

Comment:
Does this result in cpu busy looping?

Reply:
Actually it is and I didn't notice. I'll rework that part.

Reply:
I did several improvements and fixed the busy waiting. I tried several methods without much success, but I managed to get it working for all refresh rates while keeping the CPU mostly idle.

Reply:
What methods did you try that did not success?
How will the new solution work if you have multiple monitors and/or move the window around between the two monitors? What happens if the refresh rate of the monitor is changed while the program is running?

Reply:
Did you check out how I did it for the d3d12 renderer? Are any of those techniques possible with this d3d11 renderer + Windows 7 or are those APIs Windows 10 only?

Reply:
Everything I found regarding having fixed step game loops without busy waiting, relied on timers or sleeps. I tried several windows APIs for timers, along with some functions that supposedly allowed to have better precision, but they were never precise enough, so I had pretty inconsistent framerates or stuttering, or going too fast.

As long as both monitors use the same refresh rate, it works fine. But when moving for example from a monitor at 60hz to another one at 75hz, the game starts going faster, because it enabled vsync in the first monitor. And changing the refresh rate from 75 to 60 while the game is running, the game just goes to normal speed again. I haven't checked if there are some APIs to detect a window changing monitors or refresh rate changes. But if there are, it would probably easy to fix. Also, I just noticed that entering full screen in d3d12 in the other monitor (the 75hz one) it crashes. d3d11 keeps working.

I checked the d3d12 renderer. I applied some things, like drawing frames from WM_PAINT but I was getting an annoying flickering when moving or resizing the window. And even though I didn't check everything, a lot of stuff there is indeed exclusive to d3d12, or newer DXGI versions, or simply not available in Windows 7.

Reply:
To be more specific, the game applies vsync if the monitor runs at 60, 90 or 120hz. For other frequencies, it resorts to manual timing. The game speed changes only occur if the game started with vsync enabled (due to being executed in a monitor at 60, 90 or 120hz), and refresh rate changes to a different one. If the game started in manual timing mode, it will keep working fine whatever the refresh rate.

Reply:
Can this framerate detection be updated more often (like every frame, or every 10 frames)? This would make it work even if the frame rate is changed.

Reply:
Probably. First I'll see if there's a way to detect refresh rate changes. If not, yeah, it can be queried at a certain interval.

Reply:
Implemented. In the end, it's not necessary to query refresh rate changes periodically. WM_EXITSIZEMOVE and WM_DISPLAYCHANGE events cover all the cases I can think of: moving through windows, scaling the window into another monitor, changing resolution and refresh rate settings in Windows, etc. And the game's framerate stays always consistent.

Reply:
It seems this does not cover every case. Windows seems to apply a global refresh rate among all monitors, even if each of them has an individual refresh rate. What the system does is that it synchronizes to the default monitor's refresh rate, and in some cases to the one containing windows where "things happen". So when I test your code on a setup with one 60 Hz and one 120 Hz monitor, where the 60 Hz one is selected as primary, the program runs at half speed. At another test, when I have a 60 Hz as primary and a 30 Hz as secondary, the program runs fine when the window is in front on the 30 Hz screen, but as soon as all pixels is covered by another window in front, the program runs at double speed (you can here the audio playing too fast).

When "lower_latency" is available, or really when the Windows version supports it, you should use the method I use in the d3d12 code to get the correct framerate. And yeah we should fix so that it falls back to timer if it detects a refresh rate > 120 Hz.